code/controllers/subsystem/ipintel.dm

```diff
@@ -1,13 +1,273 @@
+		qdel(query)
+		return null
+
+	query.NextRow()
+	var/list/data = query.item
+	qdel(query)
+	if(isnull(data))
+		return null
+
+	var/datum/ip_intel/intel = new
+	intel.query_status = "cached"
+	intel.result = data["intel"]
+	intel.date = data["date"]
+	intel.address = address
+	return TRUE
+
+/datum/controller/subsystem/ipintel/proc/is_exempt(client/player)
+	if(exempt_living_playtime > 0)
+		var/list/play_records = player.prefs.exp
+		if (!play_records.len)
+			player.set_exp_from_db()
+			play_records = player.prefs.exp
+		if(length(play_records) && play_records[EXP_TYPE_LIVING] > exempt_living_playtime)
+			return TRUE
+	return FALSE
+
+/datum/controller/subsystem/ipintel/proc/is_whitelisted(ckey)
+	var/datum/db_query/query = SSdbcore.NewQuery(
+		"SELECT * FROM [format_table_name("ipintel_whitelist")] WHERE ckey = :ckey", list(
+			"ckey" = ckey
+		)
+	)
+	query.warn_execute()
+	query.sync()
+	if(query.status == DB_QUERY_BROKEN)
+		qdel(query)
+		return FALSE
+	query.NextRow()
+	return !!query.item // if they have a row, they are whitelisted
+
+ADMIN_VERB(ipintel_allow, R_BAN, "Whitelist Player VPN", "Allow a player to connect even if they are using a VPN.", ADMIN_CATEGORY_IPINTEL, ckey as text)
+	if(SSipintel.is_whitelisted(ckey))
+		to_chat(user, "Player is already whitelisted.")
+		return
+
+	var/datum/db_query/query = SSdbcore.NewQuery(
+		"INSERT INTO [format_table_name("ipintel_whitelist")] ( \
+			ckey, \
+			admin_ckey \
+		) VALUES ( \
+			:ckey, \
+			:admin_ckey \
+		)", list(
+			"ckey" = ckey,
+			"admin_ckey" = user.ckey,
+		)
+	)
+	query.warn_execute()
+	query.sync()
+	qdel(query)
+	message_admins("IPINTEL: [key_name_admin(user)] has whitelisted '[ckey]'")
+
+ADMIN_VERB(ipintel_revoke, R_BAN, "Revoke Player VPN Whitelist", "Revoke a player's VPN whitelist.", ADMIN_CATEGORY_IPINTEL, ckey as text)
+	if(!SSipintel.is_whitelisted(ckey))
+		to_chat(user, "Player is not whitelisted.")
+		return
+	var/datum/db_query/query = SSdbcore.NewQuery(
+		"DELETE FROM [format_table_name("ipintel_whitelist")] WHERE ckey = :ckey", list(
+			"ckey" = ckey
+		)
+	)
+	query.warn_execute()
+	query.sync()
+	qdel(query)
+	message_admins("IPINTEL: [key_name_admin(user)] has revoked the VPN whitelist for '[ckey]'")
+
+/client/proc/check_ip_intel()
+	if(SSipintel.is_exempt(src) || SSipintel.is_whitelisted(ckey))
+		return
+
+	var/intel_state = SSipintel.get_address_intel_state(address)
+	var/reject_bad_intel = CONFIG_GET(flag/ipintel_reject_bad)
+	var/reject_unknown_intel = CONFIG_GET(flag/ipintel_reject_unknown)
+	var/reject_rate_limited = CONFIG_GET(flag/ipintel_reject_rate_limited)
+
+	var/connection_rejected = FALSE
+	var/datum/ip_intel/intel = SSipintel.cached_queries[address]
+	switch(intel_state)
+		if(IPINTEL_BAD_IP)
+			log_access("IPINTEL: [ckey] was flagged as a VPN with [intel.result * 100]% likelihood.")
+			if(reject_bad_intel)
+				to_chat_immediate(src, span_boldnotice("Your connection has been detected as a VPN."))
```

Comment:
you might want to specify how they might contact the admin team for further help with getting past this point.

discord/forums maybe.


Ideally it would put them in the interview queue, but that would probably require more extensive changes.